### PR TITLE
feat: Add Method of Equal Shares

### PR DESCRIPTION
**Pull request by fsargent**
**Context:**

This pull request adds new test cases for the `MethodOfEqualShares` class within the `packages/backend/src/Tabulators/MethodOfEqualShares.ts` file. This includes importing the `fraction.js` package and several testing scenarios such as Basic Example, Influence Budget Redistribution, Random Tiebreaker, Random Tiebreaker with Defined Order, and Valid/Invalid/Under/Bullet Vote Counts.

**Key Changes:**

1. **New test file:**
   - Created a new test file at `packages/backend/src/Tabulators/MethodOfEqualShares.test.ts`.
   - Added comprehensive test cases to validate the functionality of the `MethodOfEqualShares`.

2. **Implementation of MethodOfEqualShares**
   - Added `packages/backend/src/Tabulators/MethodOfEqualShares.ts` to implement the MethodOfEqualShares function.
   - The function calculates election results based on various voting methods and scenarios.
   - It handles tiebreakers, influence budget redistribution, and validates different vote types.

**Testing Instructions:**

- Run the test suite in the `packages/backend/` directory to ensure all new and existing tests pass.

**Motivation:**

- To ensure the reliability and accuracy of the `MethodOfEqualShares` function.
- To handle edge cases and validate the workflow of the election model.

### CODE REVIEW

**Comment by fsargent** [comment]
image
```

 FAIL  src/Tabulators/MethodOfEqualShares.test.ts
  Method of Equal Shares Tests
    ✓ Basic Example (3 ms)
    ✓ Basic Example 2
    ✕ Basic Example 3 (3 ms)
    ✓ Influence Budget Redistribution
    ✓ Random Tiebreaker (1 ms)
    ✓ Random Tiebreaker with Defined Order (1 ms)
    ✓ Valid/Invalid/Under/Bullet Vote Counts (1 ms)

  ● Method of Equal Shares Tests › Basic Example 3

    expect(received).toBe(expected) // Object.is equality

    Expected: "Grace"
    Received: "Frank"

      81 |     expect(results.elected[0].name).toBe("Allison");
      82 |     expect(results.elected[1].name).toBe("Bill");
    > 83 |     expect(results.elected[2].name).toBe("Grace");
         |                                     ^
      84 |     expect(results.elected[3].name).toBe("Jack");
      85 |     expect(results.elected[4].name).toBe("Mona");
      86 |     expect(results.elected[5].name).toBe("Oscar");

      at Object.<anonymous> (src/Tabulators/MethodOfEqualShares.test.ts:83:37)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 6 passed, 7 total
Snapshots:   0 total
Time:        2.183 s, estimated 3 s
```

Not sure why this is failing...

**Comment by mikefranze** [comment]
> image
> 
> ```
> 
>  FAIL  src/Tabulators/MethodOfEqualShares.test.ts
>   Method of Equal Shares Tests
>     ✓ Basic Example (3 ms)
>     ✓ Basic Example 2
>     ✕ Basic Example 3 (3 ms)
>     ✓ Influence Budget Redistribution
>     ✓ Random Tiebreaker (1 ms)
>     ✓ Random Tiebreaker with Defined Order (1 ms)
>     ✓ Valid/Invalid/Under/Bullet Vote Counts (1 ms)
> 
>   ● Method of Equal Shares Tests › Basic Example 3
> 
>     expect(received).toBe(expected) // Object.is equality
> 
>     Expected: "Grace"
>     Received: "Frank"
> 
>       81 |     expect(results.elected[0].name).toBe("Allison");
>       82 |     expect(results.elected[1].name).toBe("Bill");
>     > 83 |     expect(results.elected[2].name).toBe("Grace");
>          |                                     ^
>       84 |     expect(results.elected[3].name).toBe("Jack");
>       85 |     expect(results.elected[4].name).toBe("Mona");
>       86 |     expect(results.elected[5].name).toBe("Oscar");
> 
>       at Object.<anonymous> (src/Tabulators/MethodOfEqualShares.test.ts:83:37)
> 
> Test Suites: 1 failed, 1 total
> Tests:       1 failed, 6 passed, 7 total
> Snapshots:   0 total
> Time:        2.183 s, estimated 3 s
> ```
> 
> Not sure why this is failing...
> 
> ```
>  {
>       candidates: [
>         { index: 0, name: 'Allison', tieBreakOrder: 0 },
>         { index: 1, name: 'Bill', tieBreakOrder: 1 },
>         { index: 2, name: 'Carmen', tieBreakOrder: 2 },
>         { index: 3, name: 'Doug', tieBreakOrder: 3 },
>         { index: 4, name: 'Eve', tieBreakOrder: 4 },
>         { index: 5, name: 'Frank', tieBreakOrder: 5 },
>         { index: 6, name: 'Grace', tieBreakOrder: 6 },
>         { index: 7, name: 'Hank', tieBreakOrder: 7 },
>         { index: 8, name: 'Ivy', tieBreakOrder: 8 },
>         { index: 9, name: 'Jack', tieBreakOrder: 9 },
>         { index: 10, name: 'Karen', tieBreakOrder: 10 },
>         { index: 11, name: 'Leo', tieBreakOrder: 11 },
>         { index: 12, name: 'Mona', tieBreakOrder: 12 },
>         { index: 13, name: 'Nina', tieBreakOrder: 13 },
>         { index: 14, name: 'Oscar', tieBreakOrder: 14 },
>         { index: 15, name: 'Paul', tieBreakOrder: 15 },
>         { index: 16, name: 'Quinn', tieBreakOrder: 16 },
>         { index: 17, name: 'Rita', tieBreakOrder: 17 },
>         { index: 18, name: 'Sam', tieBreakOrder: 18 }
>       ],
>       totalScores: [
>         { index: 0, score: 6 },
>         { index: 1, score: 6 },
>         { index: 2, score: 4 },
>         { index: 3, score: 5 },
>         { index: 4, score: 5 },
>         { index: 5, score: 6 },
>         { index: 6, score: 7 },
>         { index: 7, score: 5 },
>         { index: 8, score: 5 },
>         { index: 9, score: 6 },
>         { index: 10, score: 2 },
>         { index: 11, score: 5 },
>         { index: 12, score: 6 },
>         { index: 13, score: 4 },
>         { index: 14, score: 6 },
>         { index: 15, score: 3 },
>         { index: 16, score: NaN },
>         { index: 17, score: NaN },
>         { index: 18, score: NaN }
>       ],
>       nValidVotes: 10,
>       nInvalidVotes: 0,
>       nUnderVotes: 0,
>       nBulletVotes: 0
>     }
> ```

@fsargent does this tool output any intermediate data that we can use to debug?

**Comment by mikefranze** [comment]
Could also check the box to use fractions instead of floats and check if we're hitting any tiebreaker scenarios.

**Comment by mikefranze** [comment]
Looks like a ties issue. In the tool you can click on the results at the bottom and get logs that includes a list of tied winner sets.

Your winners were: 0, 1, 5, 6, 9, 12

Tied winner sets:
0,1,5,6,12,14
0,1,6,9,12,14
0,1,5,6,8,12
0,1,5,6,9,12 <===
0,1,5,6,9,14
0,5,6,8,12,14
0,5,6,9,12,14
0,5,6,8,9,12
1,5,6,9,12,14

File: packages/backend/src/Tabulators/MethodOfEqualShares.ts [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+    tieBreakOrder.push(...groupOrder);
+  }
+
+  return tieBreakOrder;
+}
+
+function shuffle<T>(array: T[]): T[] {
+  const shuffled = array.slice();
+  for (let i = shuffled.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
```
Comment (mikefranze) [inline packages/backend/src/Tabulators/MethodOfEqualShares.ts:547]: We shouldn't need a random generator here. The "randomTiebreakOrder" input to the main function is intended to determine how "random" ties are broken. This allows tests to be determinate and the server can generate a random order based on candidate UUIDs. Sorry that this wasn't documented.

Comment (fsargent) [reply packages/backend/src/Tabulators/MethodOfEqualShares.ts:547]: @mikefranze Yeah, makes sense. I'm having issues because the existing tool (https://pref.tools/abcvoting/) doesn't actually describe what order the candidates are elected, so it's hard for me to write tests without me just 'guessing' who should win.
